src/swap/saga.ts

```diff
@@ -90,9 +114,23 @@ export function* swapSubmitSaga(action: PayloadAction<SwapInfo>) {
   const amount = action.payload.unvalidatedSwapTransaction[amountType]
   const { quoteReceivedAt } = action.payload
 
-  const tokenBalances: TokenBalance[] = yield* select(swappableTokensSelector)
-  const fromToken = tokenBalances.find((token) => token.address === sellTokenAddress)
-  const toToken = tokenBalances.find((token) => token.address === buyTokenAddress)
+  const tokens = yield* select((state) =>
+    tokensByIdSelector(state, [networkConfig.defaultNetworkId])
+  )
+  const fromToken =
+    tokens[
+      getTokenId(
+        networkConfig.defaultNetworkId,
+        sellTokenAddress === nativeTokenAddressForSwapProviders ? undefined : sellTokenAddress
+      )
+    ]
+  const toToken =
+    tokens[
+      getTokenId(
+        networkConfig.defaultNetworkId,
+        buyTokenAddress === nativeTokenAddressForSwapProviders ? undefined : buyTokenAddress
+      )
+    ]
```

Comment:
Could/should we use `action.payload.userInput.{from|to}Token` to avoid having to know about `0xeeeee...` here?

Reply:
actually this is a great question - i wondered the same, but i started wondering if it is a risk? e.g. if the token addresses are incorrect in the response, then we potentially execute something different to what we display? maybe this is not something we should worry about?

Reply:
I'm asking myself the same question for the viem flow.

Reply:
I think we should rely on the backend response.
Using the UI data as source of truth sounds risky.

Reply:
I guess since the real thing happening is with the `data` field from the transaction, the risk of executing something different that what is displayed is there no matter if we use the user input or the `sellTokenAddress` or `buyTokenAddress`, or not.

Reply:
I actually disagree with using the (swaps) backend as a source of truth here. The swaps cloud functions are centered around the providers and 0x and openocean use different addresses for celo native tokens.

Reply:
after thinking about it some more, i tend to agree that we should trust what comes from the backend. it simplifies this swaps logic but also more broadly it is a rabbit hole to start doubting the responses that our services provide... we should indeed validate the `data` field if we want to deal with the security topic but that would require our swaps providers to verify their contracts (i have requested this from the 0x team but didn't get a response :( ) since the buy and sell token addresses from the response here does not affect the outcome of the swap, i'll make the change to not read them